connexion/operation.py

```diff
@@ -89,10 +87,34 @@ def __init__(self, method, path, operation, app_produces, app_security,
         self.security = operation.get('security', app_security)
         self.__undecorated_function = resolver(self.operation_id)
 
-    def detect_controller(self, operation_id, router_controller):
-        if router_controller is None:
+    def detect_controller(self):
+
+        operation_id = self.operation.get('operationId')
+        x_router_controller = self.operation.get('x-swagger-router-controller')
+
+        if operation_id:
+            if x_router_controller:
+                return x_router_controller + '.' + operation_id
             return operation_id
-        return router_controller + '.' + operation_id
+
+        function = self.method.lower()
+
+        if x_router_controller:
+            mod_name = x_router_controller
+        else:
+            if not self.default_module_name:
+                raise InvalidSpecification(
+                    "Neither operationId or x-swagger-router-controller was "
+                    + "configured and no default module set for Api"
+                )
+
+            mod_name = self.default_module_name
+            match = re.search('^/(.+?)(/.?|$)', self.path)
+            if match:
+                mod_name += '.' + match.group(1)
+                function = ('search' if function == 'get' and match.group(2).strip('/') == '' else function)
```

Comment:
@hjacobs What if the resolver returned both the function *and* the operationId? Then the whole algorithm would become pluggable by passing the resolver to `add_api()` as is already supported in master. `detect_controller` would become obsolete (as it would be part of the default resolver), the default, purely `operationId` based algorithm could stay the default, *but* you could easily replace it with this more 'automagic' algorithm. Eg:

```python
app.add_api('swagger.yaml', default_module_name='api', resolver=connexion.resolvers.resty)
```